src/adapter/src/coord/appends.rs

```diff
@@ -368,39 +349,34 @@ impl Coordinator {
             .wall_time()
             .observe(histogram);
 
-        if should_block {
-            // We may panic here if the storage controller has shut down, because we cannot
-            // correctly return control, nor can we simply hang here.
-            // TODO: Clean shutdown.
+        // Spawn a task to do the table writes.
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        task::spawn(|| "group_commit_apply", async move {
+            // Wait for the writes to complete.
             append_fut
                 .await
-                .expect("One-shot dropped while waiting synchronously")
+                .expect("One-shot dropped while waiting")
                 .unwrap_or_terminate("cannot fail to apply appends");
-            self.group_commit_apply(timestamp, responses, write_lock_guard, notifies, permit)
-                .await;
-        } else {
-            let internal_cmd_tx = self.internal_cmd_tx.clone();
-            task::spawn(
-                || "group_commit_apply",
-                async move {
-                    if let Ok(response) = append_fut.await {
-                        response.unwrap_or_terminate("cannot fail to apply appends");
-                        if let Err(e) = internal_cmd_tx.send(Message::GroupCommitApply(
-                            timestamp,
-                            responses,
-                            write_lock_guard,
-                            notifies,
-                            permit,
-                        )) {
-                            warn!("Server closed with non-responded writes, {e}");
-                        }
-                    } else {
-                        warn!("Writer terminated with writes in indefinite state");
-                    }
-                }
-                .instrument(Span::current()),
-            );
-        }
+
+            // Note: while technically we should have `GroupCommitApply` update the notifies, we
+            // know that internal commands get processed before any user commands, so the writes
+            // are still guaranteed to be observable before any user commands.
+            for notify in notifies {
+                // We don't care if the listeners have gone away.
+                let _ = notify.send(());
+            }
+
+            // Trigger a GroupCommitApply, which will run before any user commands since we're
+            // sending it on the internal command sender.
```

Comment:
We can't send the notifies after we apply the timestamp because that requires waiting for the Coordinator to run the queued `GroupCommitApply` command, which it won't be able to do because the current command is waiting for the notify to resolve, so we end up in a dead lock.

Joe and I talked in a Huddle, the correctness property here relies on internal commands being run before user commands. We queue a `GroupCommitApply` on the internal command sender before notifying any waiters, so we know the write will be applied before any user commands could observe the intermediate state.

We don't _love_ relying on this property though, and conveniently some PlatformV2 work will soon make it possible to give an owned handle to the timestmap oracle, to this spawned task. Then we could explicitly apply the write in the task instead of relying on the queued command to get processed before user commands. Waiting to hear back on how soon the shareable timestamp oracle will be ready!